Introduce standard devicetree Kconfig symbols for drivers

galak opened the pull request:
The intent of this PR is to standardize and hopefully ease some of the boilerplate around driver Kconfig and devicetree.

We already have a number of drivers that utilize the following pattern:

```
DT_COMPAT_NXP_IMX_CSI := nxp,imx-csi

config VIDEO_MCUX_CSI
   bool "NXP MCUX CMOS Sensor Interface (CSI) driver"
   default $(dt_compat_enabled,$(DT_COMPAT_NXP_IMX_CSI))
   ...
```

The idea is to introduce drivers/Kconfig.dts that a script can generate by scanning the source tree for all dts bindings and build a list of `compatible` values from.  The above example than would look like:

```
DT_COMPAT_NXP_IMX_CSI := nxp,imx-csi

config DTS_HAS_NXP_IMX_CSI
        def_bool $(dt_has_compat,$(DT_COMPAT_NXP_IMX_CSI))

config DTS_HAS_NXP_IMX_CSI_ENABLED
        depends on DTS_HAS_NXP_IMX_CSI
        def_bool $(dt_compat_enabled,$(DT_COMPAT_NXP_IMX_CSI))
```

and the `drivers/video/Kconfig.mcux_csi` looks like:

```
config VIDEO_MCUX_CSI
	bool "NXP MCUX CMOS Sensor Interface (CSI) driver"
	default DTS_HAS_NXP_IMX_CSI_ENABLED
	depends on DTS_HAS_NXP_IMX_CSI
	...  
```

* We introduce `DTS_HAS_<sanitized compat>_ENABLED` to make it a little easier on driver writers and to catch errors like: `default $(dt_compat_enabled,${DT_COMPAT_FOOBAR})` (use of `{`,`}`  is incorrect.
* We introduce `DT_HAS_<sanitized compat>` so that only drivers that exist in a board's devicetree are exposed (for those drivers that are devicetree based).

Additionally by introducing standardize and generated Kconfig symbols for devicetree compatibles we create an association between devicetree/Kconfig/CMake that we can utilize in the future for various purposes like doc generation or testing.

galak committed:
dts: Add simple script to generate Kconfig.dts based on bindings

Given binding dirs the script will generate a Kconfig.dts of the
form:

    DT_COMPAT_ADI_ADT7420 := adi,adt7420

    config DT_HAS_ADI_ADT7420_ENABLED
            depends on DTS_HAS_ADI_ADT7420
            def_bool $(dt_compat_enabled,$(DT_COMPAT_ADI_ADT7420))

Than a driver Kconfig can use these Kconfig symbols as follows:

    menuconfig ADT7420
            bool "ADT7420 Temperature Sensor"
            default y
            depends on DT_HAS_ADI_ADT7420_ENABLED
            ...

Signed-off-by: Kumar Gala <galak@kernel.org>

galak committed:
drivers: video: Update drivers to use devicetree Kconfig symbol

Update video drivers to use DT_HAS_<compat>_ENABLED Kconfig symbol
to expose the driver and enable it by default based on devicetree.

Signed-off-by: Kumar Gala <galak@kernel.org>

galak committed:
tests: build_all: video: remove unnecessary Kconfig symbols

Kconfig will default to enable the drivers based on devicetree so
we don't need to explicitly set the Kconfig symbols.

Signed-off-by: Kumar Gala <galak@kernel.org>

galak commented:
NOTE: This PR is based on https://github.com/zephyrproject-rtos/zephyr/pull/47158 to show the benefit for something like a build_all test in which just adding having the devicetree node added to the test gets the driver enabled and building.

carlescufi commented:
API meeting:

- @carlescufi this will add lots of needless Kconfig options to document
- @gmarull mentions that this will not be compatible with OoT drivers
- @erwango mentions that it might be a bit burdensome: add the compat to the list, regenerate the Kconfig.dts file, etc.
- The list file will likely have conflicts often
- Caspar: Could this be done at build time based on your own compatibles? @nashif does not like the idea
- @nashif alternative: Use a Kconfig template but without autogenerating anything, just reducing the boilerplate

galak commented:
> * @carlescufi this will add lots of needless Kconfig options to document

It doesn't look like `def_bool`s show in the Kconfig generated docs today?
 
> * @gmarull mentions that this will not be compatible with OoT drivers

Why OoT drivers do or don't in this area shouldn't impact things.  The drivers are OoT.  
 
> * @erwango mentions that it might be a bit burdensome: add the compat to the list, regenerate the Kconfig.dts file, etc.
> 
> * The list file will likely have conflicts often

1. The compat list would need to exist, one script will scan bindings dir and generate Kconfig.dts
2. Wonder how often we'll have conflicts if the file is sorted by name.  Likely new changes will not be next to each other.

> * Caspar: Could this be done at build time based on your own compatibles? @nashif does not like the idea
> 
> * @nashif alternative: Use a Kconfig template but without autogenerating anything, just reducing the boilerplate

henrikbrixandersen commented:
> Why OoT drivers do or don't in this area shouldn't impact things. The drivers are OoT.

But OOT drivers need to be able to use the same infrastructure as used for in-tree drivers.

gmarull commented:
In general I like the idea of standardizing these options, but my main concern is the autogenerated file. Ideas/comments:

- OOT drivers need to be able to use the same Kconfig patterns
- Offline file means more maintenance (e.g. extra CI checks, potential merge conflicts, etc.)
- I guess generating such file at runtime takes too much time?
- We could generate the file at runtime based on parsed DTS (smaller I guess) and handle non-existing options that have the `DT_...` pattern as `n` in Kconfig scripts.

erwango commented:
On my side, I'm affraid this might be cryptic to new comers: 
How to guess that a script should be run after a new binding is added?

galak committed:
dts: Introduce Kconfig.dts

Introduce a generated Kconfig.dts that sets a Kconfig symbol for
every compatible.

* We set DT_HAS_<compat>_ENABLED if the devicetree has a node with
  <compat> enabled. (status is okay...)

We can then use the Kconfig symbol in driver Kconfig to determine
if the driver should be available, and thus enabled by default.

Signed-off-by: Kumar Gala <galak@kernel.org>

galak commented:
I've updated this to now be handled at runtime.  So this should work for OoT and not require any manual intervention.

@gmarull, @henrikbrixandersen, @erwango take a look.

gmarull commented on scripts/dts/gen_driver_kconfig_dts.py:
```diff
@@ -0,0 +1,108 @@
+import yaml
+try:
+    # Use the C LibYAML parser if available, rather than the Python parser.
+    # This makes e.g. gen_defines.py more than twice as fast.
```
this message doesn't seem relevant here?

galak replied:
will fix, copy/pasted from gen_defines.py ;)

gmarull commented on scripts/dts/gen_driver_kconfig_dts.py:
```diff
@@ -0,0 +1,108 @@
+
+    binding_paths = []
+
+    for bindings_dir in bindings_dirs:
+        for root, _, filenames in os.walk(bindings_dir):
+            for filename in filenames:
+                if filename.endswith(".yaml") or filename.endswith(".yml"):
+                    binding_paths.append(os.path.join(root, filename))
```
suggestion: pathlib glob could likely simplify this (I think)

galak replied:
This is pretty much lifted from edtlib.py so going to leave it alone for now.

gmarull commented on cmake/modules/dts.cmake:
```diff
@@ -233,6 +238,18 @@ if(SUPPORTS_DTS)
+    --bindings-dirs ${DTS_ROOT_BINDINGS}
+    WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+    RESULT_VARIABLE ret
+    )
```
Maybe add a status message? Configure step is already cluttered enough, though...

galak replied:
Yeah, don't know what the status message would say/add and how much anyone pays any attention to them.

gmarull commented on cmake/modules/dts.cmake:
```diff
@@ -233,6 +238,18 @@ if(SUPPORTS_DTS)
 
+
+  execute_process(
+    COMMAND ${PYTHON_EXECUTABLE} ${GEN_DRIVER_KCONFIG_SCRIPT}
```
question: we are forcing to re-run configure stage if we add a new compatible, is it like this now?

gmarull reviewed (commented):
LGTM, a few questions/suggestions. I've tested the changes and I didn't notice a build time increase.

Btw, I've tested OOT case, and it seems to be working well, which is great:
https://github.com/zephyrproject-rtos/example-application/pull/20

galak committed:
doc: generate dummy Kconfig.dts

Generate a dummy Kconfig.dts to allow docs to build

Signed-off-by: Kumar Gala <galak@kernel.org>

galak committed:
scripts: check_compliance: generate Kconfig.dts

Generate a Kconfig.dts to allow check_compliance to run.

Signed-off-by: Kumar Gala <galak@kernel.org>

mbolivar-nordic reviewed (commented):
I think this is a great idea. Have you given any thought to how you want to document it?

galak commented:
> I think this is a great idea. Have you given any thought to how you want to document it?

What aspect of docs?   I haven't yet given that much thought.

galak commented:
One question I have is related to keeping a "cached" copy of the in `dts/Kconfig.dts`. 

Positives:
* reduces execution time of script
* reduces how big the generated file is, so a bit easier to read
* Allows for "editor" completion of Kconfig symbols.
* Able to see some level of what changed, got added between updates

Negatives:
* requires update from time to time (maybe just before or after a release) of `dts/Kconfig.dts`

galak commented:
@henrikbrixandersen, @carlescufi , @MaureenHelm , @nashif, @erwango, @gmarull, @mbolivar-nordic what else needs to be done here?

I know @mbolivar-nordic mentioned docs, so looking for peoples input on what doc updates we should make for this.

Anything else?

galak commented:
> I can't find anywhere to put this documentation, which makes me sad. One more bit of proof that our driver documentation is missing a lot of real-world details.

Agreed, was thinking how we don't have a "how to write a driver" guide.

galak commented:
> @henrikbrixandersen, @carlescufi , @MaureenHelm , @nashif, @erwango, @gmarull, @mbolivar-nordic what else needs to be done here?

Updated to just have the `DT_<COMPAT>_ENABLED` symbol.  Its all we actually need.  It doesn't make sense to allow a driver to be compiled in if at least one instance isn't enabled.  Otherwise we'll just get compile errors (and this addresses an issue with singleton devices that use `DEVICE_DT_DEFINE`.

microbuilder commented:
> > I can't find anywhere to put this documentation, which makes me sad. One more bit of proof that our driver documentation is missing a lot of real-world details.
> 
> Agreed, was thinking how we don't have a "how to write a driver" guide.

I wouldn’t mind working on that if you and @mbolivar-nordic are OK to collaborate on the DT side.

henrikbrixandersen commented on drivers/video/Kconfig.mcux_csi:
```diff
@@ -3,9 +3,8 @@
 config VIDEO_MCUX_CSI
 	bool "NXP MCUX CMOS Sensor Interface (CSI) driver"
-	default $(dt_compat_enabled,$(DT_COMPAT_NXP_IMX_CSI))
+	default DTS_HAS_NXP_IMX_CSI_ENABLED
```
How come `DTS_` and not `DT_` like we use in other devicetree APIs?

galak replied:
No real reason, I changed it to `DT_`

henrikbrixandersen reviewed (commented):
Looks really good, one small question:

henrikbrixandersen reviewed (dismissed):
Looks good, thank you.

stephanosio reviewed (dismissed):
LGTM

pdgendt commented on drivers/video/Kconfig.mt9m114:
```diff
@@ -6,5 +6,7 @@
 	bool "MT9M114 Aptina CMOS digital image sensor"
 	depends on I2C
+	depends on DT_HAS_APTINA_MT9M114_ENABLED
+	default DT_HAS_APTINA_MT9M114_ENABLED
```
What is the difference with or advantage over:
```suggestion
	default y
```

galak replied:
@pdgendt there isn't one.  It made a bit more sense before I added the `depends on`.  But with that, it should just be `default y`.

MaureenHelm reviewed (changes requested):
Thanks for cleaning this up @galak!
